### DISCUSSION TITLE
Why is the API of Parcel 2 more restrictive compared to V1?

### DISCUSSION THREAD

**Discussion by AndyOGo** [parcel-bundler/parcel#7743] [General]
First, Parcel 2 works great and is super easy when used as a CLI tool.

Though as soon as I want to use it programmatically, I'm faced with certain restrictions not present in Parcel V1.
- accessing the e.g. bundler's `options` is prohibited because almost all members are private in Parcel V2
   https://github.com/parcel-bundler/parcel/blob/47379bf8fabeb2cfe03ade8802d942388b153e5b/packages/core/core/src/Parcel.js#L80-L82
   Parcel V1 allowed to access them, like:
   https://github.com/parcel-bundler/parcel/blob/e11f0852e30fdac9ecae22398d1cf232b5eab9a2/packages/core/parcel-bundler/src/Server.js#L35-L36
- adding or testing local plugins programmatically need a separate package for each plugin #4321
- Extending / inheriting from existing plugins in Parcel 2 #7680
- using Parcel as middleware has been removed #4612 

IMHO these restrictive changes are daunting and frustrating to work with.
Could you please explain why the API is designed that way? What are the benefits?

**Comment by mischnic** [comment]
> using Parcel as middleware has been removed

This is because the server is just a regular plugin, so there's no special way it's exposed in the core JS api.

> adding or testing local plugins programmatically need a separate package for each plugin

The original idea was described here: https://github.com/parcel-bundler/parcel/issues/4321#issuecomment-603590103

> accessing the e.g. bundler's options is prohibited because almost all members are private in Parcel V2

Maybe we could expose that similar to how `options` is passed to plugins...

**Reply by AndyOGo** [reply]
Thank you for your reply @mischnic 

> This is because the server is just a regular plugin, so there's no special way it's exposed in the core JS api.

I understand.
It would be great, if I could interact to some extend also from the outside via the API with the bundler.
Or at least have the meta data available as read-only members.

> The original idea was described here: https://github.com/parcel-bundler/parcel/issues/4321#issuecomment-603590103

I see the community viewpoint behind it, though I think it puts too much control and is more a stop than a go.

> Maybe we could expose that similar to how options is passed to plugins...

It would be really nice for my case, so that I can pass a bundler object around, and other tools can read e.g. the `outDir` options or whatever they need.
